### PR TITLE
[Graph] Stream-Ordered Memory Allocation

### PR DESCRIPTION
**Pull request by liujuncheng**

### CODE REVIEW

**Comment by ouyangyu** [comment]
### ResNet50正确性测试结果

> 1. 表头`oneflow@b65b7c8_1`表示：`ONEFLOW_GRAPH_ENABLE_STREAM_ORDERED_MEMORY_ALLOCATION=1`
> 2. 表头`oneflow@b65b7c8_0`表示：`ONEFLOW_GRAPH_ENABLE_STREAM_ORDERED_MEMORY_ALLOCATION=0`



|                NVIDIA GeForce RTX 3080 Ti                 | [oneflow@b65b7c8_1](https://github.com/Oneflow-Inc/oneflow/commit/b65b7c8)  + [models@fc7cbf8d](https://github.com/Oneflow-Inc/models/commit/fc7cbf8d) | [oneflow@b65b7c8_0](https://github.com/Oneflow-Inc/oneflow/commit/b65b7c8)  + [models@fc7cbf8d](https://github.com/Oneflow-Inc/models/commit/fc7cbf8d) |
| :-------------------------------------------------------: | :----------------------------------------------------------: | :----------------------------------------------------------: |
| resnet50_graph_realdata_DCgpu_FP16_mb160_gb1280_acc1_1n8g | [8544-8990] MiB / [8844.96](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/./3080TI/ResNet50/b65b7c8_1/1n8g/resnet50_graph_realdata_DCgpu_FP16_mb160_gb1280_acc1_1n8g_b65b7c8_20230207_073850016673371.log) / 76.9 | [8550-8838] MiB / [8907.68](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/./3080TI/ResNet50/b65b7c8_0/1n8g/resnet50_graph_realdata_DCgpu_FP16_mb160_gb1280_acc1_1n8g_b65b7c8_20230207_152946514938175.log) / 76.93 |
| resnet50_graph_realdata_DCgpu_FP16_mb40_gb1280_acc4_1n8g  | [8576-8922] MiB / [8814.08](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/./3080TI/ResNet50/b65b7c8_1/1n8g/resnet50_graph_realdata_DCgpu_FP16_mb40_gb1280_acc4_1n8g_b65b7c8_20230207_095334506677562.log) / 76.94 | [8554-8814] MiB / [8922.24](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/./3080TI/ResNet50/b65b7c8_0/1n8g/resnet50_graph_realdata_DCgpu_FP16_mb40_gb1280_acc4_1n8g_b65b7c8_20230207_174230832605326.log) / 77.24 |

- 正确性无误

**Comment by ouyangyu** [comment]
### BERT正确性测试结果

> 1. 表头`oneflow@b65b7c8_1`表示：`ONEFLOW_GRAPH_ENABLE_STREAM_ORDERED_MEMORY_ALLOCATION=1`
> 2. 表头`oneflow@b65b7c8_0`表示：`ONEFLOW_GRAPH_ENABLE_STREAM_ORDERED_MEMORY_ALLOCATION=0`

|                   NVIDIA GeForce RTX 3090                    | [oneflow@0cc4a56b](https://github.com/Oneflow-Inc/oneflow/commit/0cc4a56b)  + [libai@0bdae6c6](https://github.com/Oneflow-Inc/models/commit/0bdae6c6) | [oneflow@b65b7c8_0](https://github.com/Oneflow-Inc/oneflow/commit/b65b7c8_0)  + [libai@0bdae6c6](https://github.com/Oneflow-Inc/models/commit/0bdae6c6) | [oneflow@b65b7c8_1](https://github.com/Oneflow-Inc/oneflow/commit/b65b7c8_1)  + [libai@0bdae6c6](https://github.com/Oneflow-Inc/models/commit/0bdae6c6) |
| :----------------------------------------------------------: | :----------------------------------------------------------: | :----------------------------------------------------------: | :----------------------------------------------------------: |
| libai_bert_large_pretrain_graph<br />nl24_nah16_hs1024_fp16_actrue<br />dp1_mp1_pp1_zerofalse_stage0<br />mbs32_gbs128_acc4_1n1g | [7953 MiB](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_3090/0cc4a56b/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP1_MP1_PP1_zerofalse_stage0_mbs32_gbs128_acc4_1n1g/config.yml) / [33.97 samples/s](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_3090/0cc4a56b/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP1_MP1_PP1_zerofalse_stage0_mbs32_gbs128_acc4_1n1g/output.log) | [7953 MiB](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_3090/b65b7c8_0/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP1_MP1_PP1_zerofalse_stage0_mbs32_gbs128_acc4_1n1g/config.yml) / [33.96 samples/s](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_3090/b65b7c8_0/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP1_MP1_PP1_zerofalse_stage0_mbs32_gbs128_acc4_1n1g/output.log) | [8620 MiB](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_3090/b65b7c8_1/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP1_MP1_PP1_zerofalse_stage0_mbs32_gbs128_acc4_1n1g/config.yml) / [33.87 samples/s](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_3090/b65b7c8_1/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP1_MP1_PP1_zerofalse_stage0_mbs32_gbs128_acc4_1n1g/output.log) |
| libai_bert_large_pretrain_graph<br />nl24_nah16_hs1024_fp16_actrue<br />dp1_mp1_pp1_zerofalse_stage0<br />mbs32_gbs32_acc1_1n1g | [7929 MiB](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_3090/0cc4a56b/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP1_MP1_PP1_zerofalse_stage0_mbs32_gbs32_acc1_1n1g/config.yml) / [33.56 samples/s](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_3090/0cc4a56b/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP1_MP1_PP1_zerofalse_stage0_mbs32_gbs32_acc1_1n1g/output.log) | [7929 MiB](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_3090/b65b7c8_0/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP1_MP1_PP1_zerofalse_stage0_mbs32_gbs32_acc1_1n1g/config.yml) / [33.55 samples/s](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_3090/b65b7c8_0/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP1_MP1_PP1_zerofalse_stage0_mbs32_gbs32_acc1_1n1g/output.log) | [7601 MiB](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_3090/b65b7c8_1/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP1_MP1_PP1_zerofalse_stage0_mbs32_gbs32_acc1_1n1g/config.yml) / [33.61 samples/s](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_3090/b65b7c8_1/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP1_MP1_PP1_zerofalse_stage0_mbs32_gbs32_acc1_1n1g/output.log) |
| libai_bert_large_pretrain_graph<br />nl24_nah16_hs1024_fp16_actrue<br />dp16_mp1_pp1_zerofalse_stage0<br />mbs32_gbs512_acc1_2n8g | [8752 MiB](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_3090/0cc4a56b/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP16_MP1_PP1_zerofalse_stage0_mbs32_gbs512_acc1_2n8g/config.yml) / [7903.68 samples/s](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_3090/0cc4a56b/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP16_MP1_PP1_zerofalse_stage0_mbs32_gbs512_acc1_2n8g/output.log) | [8752 MiB](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_3090/b65b7c8_0/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP16_MP1_PP1_zerofalse_stage0_mbs32_gbs512_acc1_2n8g/config.yml) / [7912.32 samples/s](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_3090/b65b7c8_0/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP16_MP1_PP1_zerofalse_stage0_mbs32_gbs512_acc1_2n8g/output.log) | [8294 MiB](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_3090/b65b7c8_1/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP16_MP1_PP1_zerofalse_stage0_mbs32_gbs512_acc1_2n8g/config.yml) / [7869.28 samples/s](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_3090/b65b7c8_1/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP16_MP1_PP1_zerofalse_stage0_mbs32_gbs512_acc1_2n8g/output.log) |





|                  NVIDIA GeForce RTX 2080 Ti                  | [oneflow@0cc4a56b](https://github.com/Oneflow-Inc/oneflow/commit/0cc4a56b)  + [libai@0bdae6c6](https://github.com/Oneflow-Inc/models/commit/0bdae6c6) | [oneflow@b65b7c8_0](https://github.com/Oneflow-Inc/oneflow/commit/b65b7c8_0)  + [libai@0bdae6c6](https://github.com/Oneflow-Inc/models/commit/0bdae6c6) | [oneflow@b65b7c8_1](https://github.com/Oneflow-Inc/oneflow/commit/b65b7c8_1)  + [libai@0bdae6c6](https://github.com/Oneflow-Inc/models/commit/0bdae6c6) |
| :----------------------------------------------------------: | :----------------------------------------------------------: | :----------------------------------------------------------: | :----------------------------------------------------------: |
| libai_bert_large_pretrain_graph<br />nl24_nah16_hs1024_fp16_actrue<br />dp4_mp1_pp1_zerofalse_stage0<br />mbs32_gbs512_acc4_1n4g | [8009 MiB](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_2080_Ti/0cc4a56b/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP4_MP1_PP1_zerofalse_stage0_mbs32_gbs512_acc4_1n4g/config.yml) / [222.64 samples/s](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_2080_Ti/0cc4a56b/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP4_MP1_PP1_zerofalse_stage0_mbs32_gbs512_acc4_1n4g/output.log) | [8009 MiB](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_2080_Ti/b65b7c8_0/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP4_MP1_PP1_zerofalse_stage0_mbs32_gbs512_acc4_1n4g/config.yml) / [222.32 samples/s](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_2080_Ti/b65b7c8_0/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP4_MP1_PP1_zerofalse_stage0_mbs32_gbs512_acc4_1n4g/output.log) | [9669 MiB](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_2080_Ti/b65b7c8_1/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP4_MP1_PP1_zerofalse_stage0_mbs32_gbs512_acc4_1n4g/config.yml) / [220.2 samples/s](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneFlowAutoTest/libai/dev_graph_stream_ordered_memory_allocation/NVIDIA_GeForce_RTX_2080_Ti/b65b7c8_1/LibAI_bert_large_pretrain_graph_nl24_nah16_hs1024_FP16_actrue_DP4_MP1_PP1_zerofalse_stage0_mbs32_gbs512_acc4_1n4g/output.log) |

- 单卡情况下，开启ACC+环境变量，显存增加约600M。
- 单卡情况下，不开ACC+环境变量，显存减少约300M。
- 单机四卡，开启ACC+环境变量，显存增加1600M
- 多机多卡情况下，不开ACC+环境变量，显存有所下降



- ### 各个case的loss曲线对齐

image


image


image

image

**Review by clackhan — APPROVED** [review]

File: oneflow/core/register/register.cpp [hunk 1]
```diff
@@ -71,9 +98,11 @@ void* Regst::comm_net_token() {
...
     token = comm_net_token_;
     if (token != nullptr) { return token; }
-    CHECK(main_mem_ptr() != nullptr);
-    CHECK(separated_header_mem_ptr() == nullptr);
-    token = Singleton<CommNet>::Get()->RegisterMemory(main_mem_ptr(),
+    CHECK(body_mem_ptr_ != nullptr);
+    CHECK(header_mem_ptr_ != nullptr);
+    CHECK(reinterpret_cast<char*>(header_mem_ptr_) + regst_desc_->HeaderByteSize4OneRegst()
+          == body_mem_ptr_);
+    token = Singleton<CommNet>::Get()->RegisterMemory(header_mem_ptr_,
```
Comment (leaves-zwx) [inline oneflow/core/register/register.cpp:105]: 这里以前的 main_mem_ptr 应该代表的是 body ptr 吧？这里为什么换成了 header ?

Comment (liujuncheng) [reply oneflow/core/register/register.cpp:105]: > 这里以前的 main_mem_ptr 应该代表的是 body ptr 吧？这里为什么换成了 header ?

原来main_mem_ptr 有两种情况，如果是 header 和 body 分开存储的，那么   main_mem_ptr 是 body，否则 main_mem_ptr 是 header 和 body，除了comm_net 外，header 和 body 的存储方式是不应该暴露的，所以取消了main_mem_ptr 相关的接口，将 comm_net 里面的逻辑改为一个潜规则，后续还会重构 comm_net去掉这个潜规则

File: oneflow/core/register/register_manager.cpp [hunk 2]
```diff
@@ -39,8 +40,58 @@ struct PackedChunkInfo {
...
+    cur_header_pointer = separated_header_mem_ptr;
+    cur_body_pointer = main_mem_ptr;
+  } else {
+    CHECK(separated_header_mem_ptr == nullptr);
+    cur_header_pointer = main_mem_ptr;
+    if (main_mem_ptr == nullptr) {
+      cur_body_pointer = nullptr;
+    } else {
+      cur_body_pointer =
+          main_mem_ptr + rt_regst_desc->GetSoleBlobDesc()->AlignedByteSizeOfBlobHeader();
```
Comment (leaves-zwx) [inline oneflow/core/register/register_manager.cpp:68]: 可以调用 BodyByteSize4OneRegst ？

Comment (liujuncheng) [reply oneflow/core/register/register_manager.cpp:68]: > 可以调用 HeaderByteSize4OneRegst ？

这部分代码来自原来的实现，只改动了必要的部分，其他逻辑保存不变

File: oneflow/core/register/register.cpp [hunk 3]
```diff
@@ -15,23 +15,56 @@ limitations under the License.
...
+      comm_net_token_(nullptr),
+      allocation_type_(allocation_type) {
+  sorted_blob_vec_.resize(regst_desc->lbi_num());
+}
 
 Regst::~Regst() {
   if (comm_net_token_ != nullptr) { Singleton<CommNet>::Get()->UnRegisterMemory(comm_net_token_); }
 }
 
+void Regst::Init(void* header_mem_ptr) {
```
Comment (leaves-zwx) [inline oneflow/core/register/register.cpp:39]: 这里叫 Init，但下面又叫做 ResetBodyMemPtr，感觉他们应该是对等的关系。而且这里 Init 语义很弱，看名字不能知道内部是干嘛的，是不是叫 ResetHeaderMemPtr 比较好。

Comment (liujuncheng) [reply oneflow/core/register/register.cpp:39]: > 这里叫 Init，但下面又叫做 ResetBodyMemPtr，感觉他们应该是对等的关系。而且这里 Init 语义很弱，看名字不能知道内部是干嘛的，是不是叫 ResetHeaderMemPtr 比较好。

本来就是不对等的，HeaderMemPtr 是不可以Reset的，只有 BodyMemPtr 是可以Reset的 ， Init 语义本来就是初始化使之可用，具体细节是不暴露的

File: oneflow/core/register/register_manager.cpp [hunk 4]
```diff
@@ -201,57 +257,6 @@ void RegstMgr::NewRegsts(const RegstDescProto& regst_desc_proto,
...
-                              cur_header_pointer + header_offset, nullptr));
-    } else {
-      blob_ptr.reset(new Blob(regst->regst_desc()->mem_case(), blob_desc,
-                              cur_header_pointer + header_offset, cur_body_pointer + body_offset));
-      InitNonPODTypeBlobIfNeed(Singleton<MemoryAllocator>::Get(), blob_ptr.get());
-    }
-    regst->SetBlobByOrdinal(ordinal, std::move(blob_ptr));
-    const int64_t regst_desc_id = rt_regst_desc->regst_desc_id();
-    const auto& parallel_ctx = regst_desc_id2parallel_ctx_.at(regst_desc_id);
-    if (parallel_ctx.has_parallel_id()) {
```
Comment (leaves-zwx) [inline oneflow/core/register/register_manager.cpp:245]: 这里之前为什么还需要考虑 parallel id，现在又不需要了呢？

File: oneflow/core/lazy/actor/actor.cpp [hunk 5]
```diff
@@ -422,14 +419,15 @@ int Actor::HandlerZombie(const ActorMsg& msg) {
...
+    PrepareProducedNaiveInplaceDataRegst();
     Act();
 
     AsyncSendCustomizedProducedRegstMsgToConsumer();
     AsyncSendNaiveProducedRegstMsgToConsumer();
     AsyncSendInplaceProducedRegstMsgToConsumer();
 
-    AsyncSendCustomizedConsumedRegstMsgToProducer();
     AsyncSendNaiveConsumedRegstMsgToProducer();
+    AsyncSendCustomizedConsumedRegstMsgToProducer();
```
Comment (leaves-zwx) [inline oneflow/core/lazy/actor/actor.cpp:430]: 这里顺序还是先向下游发送消息，再向上游发送消息？

File: oneflow/core/register/register_manager.cpp (hunk 4 above)
Comment (liujuncheng) [reply oneflow/core/register/register_manager.cpp:245]: > 这里之前为什么还需要考虑 parallel id，现在又不需要了呢？

dead code

File: oneflow/core/lazy/actor/actor.cpp (hunk 5 above)
Comment (liujuncheng) [reply oneflow/core/lazy/actor/actor.cpp:430]: > 这里顺序还是先向下游发送消息，再向上游发送消息？

actor 目前的实现里面有错误的耦合，不能调整

**Review by leaves-zwx — APPROVED** [review]

**Pull request merged**